docs/blog/2021-08-31-seeding-databases-tutorial.md

```diff
@@ -0,0 +1,733 @@
+
+Everything that goes into a schema file is a [schema node][synth-schema]. Schema
+nodes can be identified by the `"type"` field which specifies which kind of node
+it is. The documentation pages have
+a [complete taxonomy of schema nodes][synth-generators] and their `"type"`.
+
+### Generating ids
+
+Let's look back at our [`User` model](#prisma-is-awesome). It has four
+properties:
+
+- `id`
+- `createdAt`
+- `email`
+- `nickname`
+
+Let's start with `id`. How can we generate that?
+
+The type of the `id` property in the [`User` model](#prisma-is-awesome)
+is `Int`:
+
+```graphql
+  id        Int      @unique @default(autoincrement())
+``` 
+
+and the attribute indicates that the field is meant to increment sequentially,
+going through values 0, 1, 2 etc.
+
+The `synth` schema type for numbers is [`number`][synth-number].
+Within [`number`][synth-number] there are three varieties of generators:
+
+- [`range`][synth-range]
+- [`constant`][synth-constant]
+- [`id`][synth-id]
+
+What decides the variant is the presence of the `"range"`, `"constant"`or `"id"`
+field in the node's specification.
+
+For example, a [`range`][synth-range] variant would look like
+
+```json synth
+{
+	"type": "number",
+	"range": {
+	    "low": 5,
+	    "high": 10,
+	    "step": 1
+	}
+}
+```
+
+whereas a [`constant`][synth-constant] variant would look like
+
+```json synth
+{
+    "type": "number",
+    "constant": 42
+}
+```
+
+For the `id` field we should use the [`id`][synth-id] variant, which is
+auto-incrementing. Here is an example of [`id`][synth-id] used in an array so we
+can see it behaves as expected:
+
+```json synth
+{
+    "type": "array",
+    "length": 10,
+    "content": {
+        "type": "number",
+        "id": {}
+    }
+}
+```
+
+### Generating emails
+
+Let us now look at the [`email`](#prisma-is-awesome) field of
+our [`User` model](#prisma-is-awesome):
+
+```graphql
+  email     String   @unique
+```
+
+Its type in the data model is that of a `String`. The `synth` schema type for
+that is [`string`][synth-string].
+
+There are many different variants of [`string`][synth-string] and they are
+all [exhaustively documented][synth-string]. The different variants are
+identified by the presence of a distinguishing field which can be
+
+- `"faker"`
+- `"pattern"`
+- `"uuid"`
+- [and a lot more][synth-string]...
+
+Since we are interested in generating email addresses, we will be using
+the [`"faker"`][synth-faker] variant which leverages a preset collection of
+generators for common properties like usernames, addresses and emails:
+
+```json synth
+{
+    "type": "string",
+    "faker": {
+        "generator": "safe_email"
+    }
+}
+```
+
+### Generating objects
+
+OK, so we now know how to generate the `id` and the `email` properties of
+our [`User` model](#prisma-is-awesome). But we do not yet know how to put them
+together in one object. For that we need the [`object`][synth-object] type:
+
+```json synth
```

Comment:
If you want to show something a little more sophisticated here, you can do something like:

```json
{
    "type": "array",
    "length": {
        "type": "number",
        "constant": 3
    },
    "content": {
        "type": "object",
        "id": {
            "type": "number",
            "id": {
				"start_at": 1
			}
        },
        "nickname": {
            "type": "string",
            "faker": {
                "generator": "username"
            }
        },
        "email": {
            "type": "string",
            "format": {
                "format": "{name}@example.com",
				"arguments": {
					"name": "@collection.content.nickname"
				}
            }
        }
    }
}

```